src/storage/src/hummock/file_cache/store.rs

```diff
@@ -12,431 +12,533 @@
+                        window_to_cache_size_ratio: foyer_store_config
+                            .lfu_window_to_cache_size_ratio,
+                        tiny_lru_capacity_ratio: foyer_store_config.lfu_tiny_lru_capacity_ratio,
+                    },
+                    device_config: DeviceConfig {
+                        dir: foyer_store_config.dir.clone(),
+                        capacity,
+                        file_capacity,
+                        align: foyer_store_config.device_align,
+                        io_size: foyer_store_config.device_io_size,
+                    },
+                    admissions,
+                    reinsertions: vec![],
+                    buffer_pool_size: foyer_store_config.buffer_pool_size,
+                    flushers: foyer_store_config.flushers,
+                    flush_rate_limit: foyer_store_config.flush_rate_limit,
+                    reclaimers: foyer_store_config.reclaimers,
+                    reclaim_rate_limit: foyer_store_config.reclaim_rate_limit,
+                    recover_concurrency: foyer_store_config.recover_concurrency,
+                    event_listeners: foyer_store_config.event_listener,
+                    prometheus_config: PrometheusConfig {
+                        registry: foyer_store_config.prometheus_registry,
+                        namespace: foyer_store_config.prometheus_namespace,
+                    },
+                    clean_region_threshold: foyer_store_config.reclaimers
+                        + foyer_store_config.reclaimers / 2,
+                };
+
+                FoyerStore::open(c).await.map_err(FileCacheError::foyer)
+            })
+            .await
+            .unwrap()?;
+
+        Ok(Self::FoyerRuntime {
+            runtime: Arc::new(runtime.into()),
+            store,
+            enable_filter,
+        })
     }
 
-    pub fn cache_file_size(&self) -> usize {
-        self.cache_file.size()
+    #[tracing::instrument(skip(self, value))]
+    pub async fn insert(&self, key: K, value: V) -> Result<bool> {
+        match self {
+            FileCache::None => Ok(false),
+            FileCache::Foyer { store, .. } => store
+                .insert_if_not_exists(key, value)
+                .await
+                .map_err(FileCacheError::foyer),
+            FileCache::FoyerRuntime { runtime, store, .. } => {
+                let store = store.clone();
+                runtime
+                    .spawn(async move { store.insert_if_not_exists(key, value).await })
+                    .await
+                    .unwrap()
+                    .map_err(FileCacheError::foyer)
+            }
+        }
     }
 
-    pub fn cache_file_len(&self) -> usize {
-        self.cache_file.len()
+    #[tracing::instrument(skip(self))]
+    pub fn insert_without_wait(&self, key: K, value: V) {
+        match self {
+            FileCache::None => {}
+            FileCache::Foyer { .. } => panic!("unsupported"),
```

Comment:
If it is unsupported, shall we assume that we always have a separate runtime for foyer?

Reply:
Yep. I've tested foyer w/wo separate runtime. The performance of foyer without a separate runtime is badly affected by risingwave. IMO, foyer will always be used with a separate runtime. So, I think I should remove `FileCache::Foyer`. 🤔